design/secure-communication-design.md

```diff
@@ -0,0 +1,88 @@
+# Secure Communication between Contour and Envoy
+
+_Status_: Draft, in review.
+
+This document outlines how to secure the communication between Contour
+and Envoy containers.
+
+## Goals
+
+- Achieve secure communication between Contour and Envoy
+- Authenticate Envoy connecting to Contour
+- Rotating and invalidating the credentials
+
+## Non-goals
+
+- Support for dynamic credentials
+- Support for use of Secret Discovery Service for credential management
+
+## Background
+
+Currently, Contour and Envoy containers are co-located within a pod.
+Current security model is relying on the fact that Contour exposes the management
+server on loopback interface which prevents any unauthorized access from outside
+the pod.
+
+With the need to move away from the [current deployment pattern of co-locating
+Contour and Envoy][0], it is necessary to secure the communication channel between
+the containers.
+
+## High level design
+
+- Use mutual TLS to secure communication channel between Contour and Envoy
+- Bootstrap phase of Contour will generate necessary TLS certificates
+
+## Detailed design
+
+### Default options for secure communication
+
+By default, mutual TLS is enabled to secure the communication between Contour and Envoy
+containers. The default `contour bootstrap` will generate the bootstrap configuration that
+enables secure communication between the containers. If `contour bootstrap` is passed with
+the parameter `--disable-tls`, the bootstrap configuration generated will make use of insecure
+communication between the containers.
+
+### Bootstrapping process
+
+This bootstrap phase is independent of `contour bootstrap` and uses `contour bootstrap-tls`.
+`contour bootstrap-tls` is run as an init container to bootstrap the TLS configuration.
+The bootstrap phase will create three pairs of the TLS certificates. The details about
```

Comment:
What are the params on the certs that are created? How long are they good for? Do we need to worry about adding DNS to the SANs of the cert. Right now when we deploy a split model, we grab the service IP from Kubernets ENV var to know how Envoy will find Contour's gRPC endpoint. If it's over tls we'll need to figure out how to connect to Contour over dns (I had issues last time doing this), and we'd need to generate proper SANs on the cert.

Reply:
Yes, I will update the design doc with more details about the cert expiration and the SANs.

And about the split model cases, I thought will be worked on as part of #881.

Reply:
As commented previously, I think we need to think about the split model in this design.

Reply:
Sure, I will consider this for this design.